File: DSA/VenikaPanwar_590013755/Day_81_Question_1.java
public class Day_81_Question_1 {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // Returns sum of values in [L, R] using pruned traversal
    public static int rangeSumBST(TreeNode root, int L, int R) {
        if (root == null) return 0;
        if (root.val < L) return rangeSumBST(root.right, L, R);
        if (root.val > R) return rangeSumBST(root.left, L, R);
        return root.val + rangeSumBST(root.left, L, R) + rangeSumBST(root.right, L, R);
    }

    public static void main(String[] args) {
        // Example 1: Tree 10 / \ 5 15 / \ \ 3 7 18, L=7, R=15
        TreeNode root1 = new TreeNode(10);
        root1.left = new TreeNode(5);
        root1.right = new TreeNode(15);
        root1.left.left = new TreeNode(3);
        root1.left.right = new TreeNode(7);
        root1.right.right = new TreeNode(18);
        System.out.println("Range Sum (7,15): " + rangeSumBST(root1, 7, 15)); // Output: 32

        // Example 2: Tree 10 / \ 5 15 / \ / \ 3 7 13 18, L=6, R=10
        TreeNode root2 = new TreeNode(10);
        root2.left = new TreeNode(5);
        root2.right = new TreeNode(15);
        root2.left.left = new TreeNode(3);
        root2.left.right = new TreeNode(7);
        root2.right.left = new TreeNode(13);
        root2.right.right = new TreeNode(18);
        System.out.println("Range Sum (6,10): " + rangeSumBST(root2, 6, 10)); // Output: 17
    }
}

File: DSA/GEHNA_KHANNA_590013140/day81question1.c
#include <stdio.h>
#include <stdlib.h>

// Definition of a BST Node
struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

// Utility function to create a new node
struct TreeNode* newNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

// Function to insert into BST
struct TreeNode* insert(struct TreeNode* root, int val) {
    if (root == NULL) return newNode(val);
    if (val < root->val) root->left = insert(root->left, val);
    else root->right = insert(root->right, val);
    return root;
}

// Function to compute range sum using pruned traversal
int rangeSumBST(struct TreeNode* root, int L, int R) {
    if (root == NULL) return 0;

    if (root->val < L) 
        return rangeSumBST(root->right, L, R);
    else if (root->val > R) 
        return rangeSumBST(root->left, L, R);
    else 
        return root->val + rangeSumBST(root->left, L, R) + rangeSumBST(root->right, L, R);
}

int main() {
    struct TreeNode* root = NULL;
    int n, val, L, R;

    printf("Enter number of nodes: ");
    scanf("%d", &n);

    printf("Enter %d values:\n", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &val);
        root = insert(root, val);
    }

    printf("Enter range L and R: ");
    scanf("%d %d", &L, &R);

    int sum = rangeSumBST(root, L, R);
    printf("Range Sum [%d, %d] = %d\n", L, R, sum);

    return 0;
}
